### ISSUE TITLE
Enable streaming references using :object encoding.

### ISSUE THREAD

**Issue by thallgren** [ohler55/oj#571]
When using  http://www.ohler.com/oj/doc/file.Encoding.html to stream data that contains references, the referenced value must include an ID (the `^i` tag) before a reference (the `^r` tag) is used. This is problematic when streaming since there is it isn't known if a streamed value will be referenced or not as the streaming progresses. Other encoding solutions solves this problem by giving each value that is streamed an implicit ID which is the value's index in the stream. Since the receiver reads the values in the same order as they were written, their implicit ID is known at the receiver end too.

Let's assume that the special tag `^R` was used to denote this implicit ID. The example in the encoding document:
```
{"^o":"Oj::Bag","^i":1,"x":["^i2",true],"me":"^r1"}
```
could then be rewritten as:
```
{"^o":"Oj::Bag","x":[true],"me":"^R0"}
```
`"^R0"` here means "object with index zero in the stream", i.e. the first `object`. If the stream were to reference the `[true]` array, it would use `"^R2"` (assuming that the index denotes the non-encoded values)

0 = the `Oj::Bag` instance
1 = the `"x"` key
2 = the `[true]` array
3 = the `true` in the `[true]` array
4 = the `"me"` key
5 = the reference

**Comment by ohler55** [comment]
You are correct, that approach would save the need for the `^i` tag. At this point I'm hesitant to change the format though as the space savings is small and it would break a lot of implementations that use object mode. I could see a new mode as a possibility though or possibly something that is backward compatible.

**Comment by thallgren** [comment]
The reason I used capital R in the example was to show that this could be an addition that doesn't affect backward compatibility. Let `"^i"` and `"^r"` remain as they are and let `"^R"` (or some other letter) introduce the new behavior.

**Comment by thallgren** [comment]
To clarify, let object mode be able to read the new format but keep the old format for serialization. I realize that a new mode is needed in order to switch how the serialization is performed.

**Comment by ohler55** [comment]
I'll keep this on the list but I'm afraid it is not a high priority right now.
